### PR TITLE
*: garbage collect useless stats info

### PR DESCRIPTION
**Pull request by lamxTyler**
Before this PR, there will be some useless stats info: 

- When dropping a table, we will keep the meta info of that table in the stats so that other tidb could know that it is deleted.

- If there are some error happens when deleting the stats info for drop index/column statement, these stats info will not be deleted anymore. 

In this PR, we garbage collect these useless stats by letting the stats owner periodically scan the stats info.

### CODE REVIEW

**Comment by hanfei1991** [comment]
We'd better only delete stats in gc logic

**Comment by lamxTyler** [comment]
PTAL @hanfei1991

File: statistics/gc.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	return nil
+}
+
+func (h *Handle) gcTableStats(is infoschema.InfoSchema, tableID int64) error {
+	sql := fmt.Sprintf("select is_index, hist_id from mysql.stats_histograms where table_id = %d", tableID)
+	rows, _, err := h.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(h.ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(rows) == 0 {
```
Comment (hanfei1991) [inline statistics/gc.go]: add comment for this line.

File: statistics/gc_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+func (s *testStatsUpdateSuite) TestGCStats(c *C) {
+	defer cleanEnv(c, s.store, s.do)
+	testKit := testkit.NewTestKit(c, s.store)
+	testKit.MustExec("use test")
+	testKit.MustExec("create table t(a int, b int, index idx(a, b))")
+	testKit.MustExec("insert into t values (1,1),(2,2),(3,3)")
+	testKit.MustExec("analyze table t")
+
+	testKit.MustExec("alter table t drop index idx")
+	c.Assert(len(testKit.MustQuery("select * from mysql.stats_histograms").Rows()), Equals, 3)
```
Comment (hanfei1991) [inline statistics/gc_test.go]: This checking way is strange. Why not query `select count(*)` ?

File: domain/domain.go [hunk 3]
```diff
@@ -645,31 +665,26 @@ func (do *Domain) updateStatsWorker(ctx context.Context, lease time.Duration) {
 			if err != nil {
 				log.Error("[stats] load histograms fail: ", errors.ErrorStack(err))
 			}
+		case <-gcStatsTicker.C:
+			if !owner.IsOwner() {
+				continue
+			}
+			err := statsHandle.GCStats(do.InfoSchema())
```
Comment (hanfei1991) [inline domain/domain.go]: Can we make sure that ddl is working well before executing gc?

Comment (lamxTyler) [reply domain/domain.go]: Why do we need to make sure that?

Comment (hanfei1991) [reply domain/domain.go]: An outdated infoschema is horrible, it may lack some table info so we think it has been deleted....

Comment (lamxTyler) [reply domain/domain.go]: I think that we can make sure that, 10 lease before `PrevLastUpdateVersion` is quite long...

**Comment by hanfei1991** [comment]
LGTM

File: domain/domain.go [hunk 4]
```diff
@@ -590,23 +590,43 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
 	if lease <= 0 {
 		return nil
 	}
+	owner := do.newStatsOwner()
 	do.wg.Add(1)
-	go do.updateStatsWorker(ctx, lease)
+	go do.updateStatsWorker(ctx, lease, owner)
```
Comment (zimulala) [inline domain/domain.go]: do has the statsLease, so we needn't pass `lease` as an argument.

File: domain/domain.go [hunk 5]
```diff
@@ -590,23 +590,43 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
 	if RunAutoAnalyze {
 		do.wg.Add(1)
-		go do.autoAnalyzeWorker(lease)
+		go do.autoAnalyzeWorker(lease, owner)
```
Comment (zimulala) [inline domain/domain.go]: Ditto.

File: statistics/gc.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/pingcap/tidb/util/sqlexec"
+	goctx "golang.org/x/net/context"
+)
+
+// GCStats will garbage collect the useless stats info. For dropped tables, we will first update their version so that
+// other tidb could know that table is deleted.
+func (h *Handle) GCStats(is infoschema.InfoSchema) error {
+	// To make sure that all the deleted tables have been acknowledged to all tidb,
+	// we only garbage collect version before 10 lease.
+	offset := oracle.ComposeTS(10*int64(h.Lease), 0)
```
Comment (zimulala) [inline statistics/gc.go]: I think the `lease` need the DDL lease, but here you use the stats lease.

**Comment by lamxTyler** [comment]
PTAL @zimulala

File: statistics/gc_test.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+
+package statistics_test
+
+import (
+	"math"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/util/testkit"
+	"time"
```
Comment (zimulala) [inline statistics/gc_test.go]: Move this to line 18.

**Pull request merged**